RFC-0005-structured-kernel-definitions.md

```diff
@@ -0,0 +1,323 @@
+ }
+```
+
+The code generator then generates the boilerplate code to put these
+functions together:
+
+```
+namespace native {
+
+Tensor upsample_nearest1d_cuda(const Tensor& self, IntArrayRef output_size, optional<double> scales) {
+  CUDADeviceGuard g(self.device());
+  TensorMeta expected = meta::upsample_nearest1d(self, output_size, scales);
+  Tensor result = native::empty(expected.sizes(), expected.options());
+  upsample_nearest1d_structured_cuda(expected, result, self, output_size, scales);
+  return result;
+}
+
+Tensor& upsample_nearest1d_out_cuda(Tensor& result, const Tensor& self, IntArrayRef output_size, optional<double> scales) {
+  // In event of multiple tensor arguments, code generation should
+  // be responsible for making sure all devices are consistent
+  CUDADeviceGuard g(self.device());
+  TensorMeta expected = meta::upsample_nearest1d(self, output_size, scales);
+  TORCH_CHECK(expected.options == result.options());
+  result.resize_(expected.size());
+  upsample_nearest1d_structured_cuda(expected, result, self, output_size, scales);
+  increment_version(result);
+  return result;
+}
+
+// CPU follows similarly
+
+}
+
+```
+
+Notice that the generated boilerplate here handles several concerns:
+
+* It properly sets up device guards for the kernel
+
+* For the `_out` variant, it tests that the inferred tensor options are
+  consistent with the passed in tensor
+
+* It centralizes the PyTorch convention that improperly sized out=
+  arguments get resized to fit the output
+
+* It handles version counter bumping for mutations
+
+* It is carefully written so that there are no extra redispatches; all
+  calls are direct non-dispatched function calls
+
+It’s also worth noting what things the boilerplate here does not handle:
+
+* It forces the allocation of a TensorMeta struct per output tensor,
+  before the final allocation of the output tensor. You could save a few
+  instructions by eliminating this struct entirely. However, I could not
+  think of an elegant API for doing so.
+
+In absolute terms, the boilerplate code saved here is not all that
+large; only several lines. However, when multiplied over the number of
+kernels in PyTorch, and the subtlety of remembering to handle all of
+these issues, and we think the use of code generation to automatically
+generate this boilerplate is worth it.
+
+The functionality code generated here is not made available to other
+backends. This is problematic, so we also generate a separate dispatch
+key handler which will automatically handle shape checking and other
+concerns for backends that are not concerned with performance:
+
+```
+// Name TBD; for this example we will call it DispatchKey::Common, as the
+// functionality here is common to all backends. This is an alias key
+// that resolves CommonXLA/CommonMSNPU/... in the same way as Autograd.
+
+Tensor upsample_nearest1d_common(const Tensor& self, IntArrayRef output_size, optional<double> scales) {
```

Comment:
We could offer a simple metaprogram for them so they don't have to manually write this for each op. But if I understand correctly, they only have to write this for ops they introduce right? Ops from native_functions.yaml that the backend only extends will already have shape checking etc. through our codegen right?

Reply:
Yes